src/core/icon.js

```diff
@@ -118,32 +135,57 @@ export class Icon {
   }
 
   /**
-   * The resource associated with this icon.
+   * The pixel ratio for which the icon creative is intended.
    *
-   * @type {Resource}
+   * Bounded: 0-1
+   *
+   * @type {?string}
+   * @since VAST 4.0
    */
-  get resource () {
-    return this._resource
+  get pxratio () {
+    return this._pxratio
   }
 
-  set resource (value) {
-    this._resource = value
+  set pxratio (value) {
+    this._pxratio = value
   }
 
+  // Sub-element(s).
+
   /**
-   * The click tracking configuration for this icon.
+   * The resources associated with this icon.
    *
-   * @type {IconClicks}
+   * Bounded: 0+
+   *
+   * @type {Resource[]}
    * @readonly
    */
-  get clicks () {
-    return this._clicks
+  get resources () {
+    return this._resources
+  }
+
+  /**
+   * The click tracking configuration for this icon.
+   *
+   * Bounded: 0-1
+   *
+   * @type {?IconClicks}
```

Comment:
Since this is now nullable, iab-vast-parser will need to create it conditionally. Are there other similar cases? As far as I can remember, I originally tried to be consistent about creating subobjects, so that it would be predictable for the user which properties were nullable. However, it's often a gray zone, so now's a good time to reevaluate that.

Reply:
I am creating an instance of IconClicks in the constructor. You won't have to allocate it in the parser but it will show up as an empty object in the output if not used.

Reply:
You're not creating it in the constructor. On [line 11](https://github.com/zentrick/iab-vast-model/pull/46/files/adbd82dc46ec514f0a419c1cfa10129317ccea7d#diff-1c91564d40cb1d38d7e684f449136018R11), you removed the implicit allocation of `.clicks` in favor of (indeed) having to set it from the parser, meaning that the property became nullable. This also explains why you added a setter. Or am I missing something?

Reply:
I had change it. And I was creating it in the constructor (hence my comment) but I think I will get back to what I had. It is probably easier to check for a null pointer than an empty object. What do you think ?

Reply:
That's definitely easier. The reasoning was that you'd never need to check for an empty object, but indeed, serialization might for example need this. Of course, an instance of `IconClicks` with no actual clicks is currently a valid object, so theoretically, a serializer might still choke on it. It's a lot like an array-typed property being nullable vs. empty: even if `array` is not null, `array.length` can still be `0` and therefore require an extra `if`. To accommodate that, perhaps it's easier to have an `.isEmpty` on any class that models a collection of things?

Reply:
What if we add following functions getAttributes(), getChildren() and getContent() for erach class. getAttributes() for instance will return an array of names/values that lists each attribute that was set. I guess this will make it very easy to work with the model.

Reply:
"Attributes" and "children" are precisely what I want to avoid. There's already a model for this; it's called the XML DOM API. The purpose of iab-vast-model is to convert those generic things into model-specific classes and properties.

Reply:
ok